File: Samples/DeviceAdminExample/app/src/main/java/com/example/user/deviceadminexample/MainActivity.java
package com.example.user.deviceadminexample;

import android.app.admin.DevicePolicyManager;
import android.content.ComponentName;
import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

public class MainActivity extends AppCompatActivity {


    private static final String TAG = "MainActivity";
    DevicePolicyManager devicePolicyManager;
    ComponentName componentName;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        componentName = new ComponentName(this, DeviceRecieverSample.class);
        devicePolicyManager = (DevicePolicyManager) getSystemService(DEVICE_POLICY_SERVICE);


    }


    //this is the button click
    public void lockme(View view) {

        Log.d(TAG, "lockme: ");
        Log.d(TAG, "lockme: " + devicePolicyManager.isAdminActive(componentName));


        if (devicePolicyManager.isAdminActive(componentName)) {
            devicePolicyManager.lockNow();
        }
        else{

            Intent intent = new Intent(DevicePolicyManager.ACTION_ADD_DEVICE_ADMIN);
            intent.putExtra(DevicePolicyManager.EXTRA_DEVICE_ADMIN,componentName);
            intent.putExtra(DevicePolicyManager.EXTRA_ADD_EXPLANATION, "Locking the device");
            startActivity(intent);

        }


    }
}

File: Samples(kotlin)/kotlin-coroutines/app/src/main/java/com/example/user/kotlin_coroutines/MainActivity.kt
package com.example.user.kotlin_coroutines

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.*

class MainActivity : AppCompatActivity() {

    var activityJob = Job()
    val activityScope = CoroutineScope(Dispatchers.Main + activityJob)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        printTime("runBlocking", this) {
            runBlocking {
                delay(DELAY_TIME)
                printThread("runBlocking", this)
            }
        }

        btnDoSomething.setOnClickListener {
            //testingGlobalScope()
            testingActivityScope()

        }

    }

    fun testingGlobalScope() {
        startSampleJob("GlobalScope", GlobalScope)
    }


    /**
     * The activity scope is going to be cancel the coroutines if the activity
     * if not on the foreground.
     *
     * Test the application by clicking the button "Do Something" and then moving
     * away from the activity. You would notice that the coroutine after the delay
     * is never executed.
     */

    private fun testingActivityScope() {

        startSampleJob("activityScope", activityScope)

    }

    /**
     * Cancelling the children on the onStop method on the activity would cancel
     * all the Coroutines running on that job. It would not cancel the job object itself
     * so the doSomething button click would be able to restart the job
     */

    override fun onStop() {
        super.onStop()
        activityJob.cancelChildren()
    }


    private fun startSampleJob(scopeName: String, scope: CoroutineScope) {
        scope.launch {
            async { customDelay(1_000) }
            customDelay(0)
            printThread("$scopeName before delay", this)//runs on the main thread

            delay(DELAY_TIME)//suspending function to cause a delay
            printThread("$scopeName default launch", this)
            withContext(Dispatchers.Unconfined) {
                printThread("withContext UnConfined", this)
            }
            withContext(Dispatchers.IO) {
                printThread("withContext IO", this)
            }
        }
    }

}
